src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php

```diff
@@ -300,6 +291,457 @@ public static function analyze(
+                            }
+
+                            foreach ($return_type->getAtomicTypes() as $sub_atomic_type) {
+                                if ($sub_atomic_type instanceof TLiteralString) {
+                                    $valid_ints[] = new TLiteralInt((int) $sub_atomic_type->value);
+                                } elseif ($sub_atomic_type instanceof TNumericString) {
+                                    $castable_types[] = new TInt();
+                                } else {
+                                    // any normal string is technically $valid_int[] = new TLiteralInt(0);
+                                    // however we cannot be certain that it's not inferred, therefore less strict
+                                    $castable_types[] = new TInt();
+                                }
+                            }
+
+                            continue 2;
+                        }
+                    }
+
+                    if ($intersection_type instanceof TObjectWithProperties
+                        && isset($intersection_type->methods['__toString'])
+                    ) {
+                        $castable_types[] = new TInt();
+
+                        continue 2;
+                    }
+                }
+            }
+
+            if ($atomic_type instanceof TNonEmptyArray
+                || $atomic_type instanceof TNonEmptyList
+            ) {
+                $risky_cast[] = $atomic_type->getId();
+
+                $valid_ints[] = new TLiteralInt(1);
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TArray
+                || $atomic_type instanceof TList
+                || $atomic_type instanceof TKeyedArray
+            ) {
+                // if type is not specific, it can be both 0 or 1, depending on whether the array has data or not
+                // welcome to off-by-one hell if that happens :-)
+                $risky_cast[] = $atomic_type->getId();
+
+                $valid_ints[] = new TLiteralInt(0);
+                $valid_ints[] = new TLiteralInt(1);
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TTemplateParam) {
+                $atomic_types = array_merge($atomic_types, $atomic_type->as->getAtomicTypes());
+
+                continue;
+            }
+
+            // always 1 for "error" cases
+            $valid_ints[] = new TLiteralInt(1);
+
+            $invalid_casts[] = $atomic_type->getId();
+        }
+
+        if ($invalid_casts) {
+            IssueBuffer::maybeAdd(
+                new InvalidCast(
+                    $invalid_casts[0] . ' cannot be cast to int',
+                    new CodeLocation($statements_analyzer->getSource(), $stmt)
+                ),
+                $statements_analyzer->getSuppressedIssues()
+            );
+        } elseif (!empty($risky_cast)) {
```

Comment:
@orklah didn't we talk about banning `empty` in the Psalm config at some point? It does seem like a rather pointless function since it basically just casts everything to `bool`. Personally I prefer an explicit `count($arr) !== 0` or `$arr !== []`, but we should probably have some sort of consistency.

Reply:
We still have a few to remove, but yeah, best not to add new ones

Reply:
I don't see it, did you forget to push that part?